test/e2e/router_status_test.go

```diff
@@ -0,0 +1,159 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	routev1 "github.com/openshift/api/route/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TestDeleteIngressControllerShouldClearRouteStatus deletes an Ingress Controller and
+// makes sure a previously admitted route has its status cleared since.
+func TestDeleteIngressControllerShouldClearRouteStatus(t *testing.T) {
+	// Create an ingress controller that can admit our route.
+	icName := types.NamespacedName{Namespace: operatorNamespace, Name: "ic-delete-test"}
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newLoadBalancerController(icName, domain)
+	ic.Spec.RouteSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"type": icName.Name,
+		},
+	}
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
```

Comment:
Make sure you delete the ingresscontroller even if `t.Fatalf` is called on line 41, line 45, or line 52.
```suggestion
	if err := kclient.Create(context.TODO(), ic); err != nil {
		t.Fatalf("failed to create ingresscontroller: %v", err)
	}
	defer func() {
		if err := kclient.Get(context.TODO(), icName, ic); err != nil {
			if apierrors.IsNotFound(err) {
				return
			}
			t.Fatalf("failed to get ingresscontroller %q: %v", icName, err)
		}
	 	assertIngressControllerDeleted(t, kclient, ic)
	}()
```
Actually, it might be worth tweaking `assertIngressControllerDeleted` to handle this case for you, but that involves a few steps (which I haven't tested):
1. Rename the `"k8s.io/apimachinery/pkg/api/errors"` import to `apierrors "k8s.io/apimachinery/pkg/api/errors"`, and update all references to the package (s/errors/apierrors).  (This step is optional—you could import the `"errors"` package under a different name—but it's more consistent with other code to rename `"k8s.io/apimachinery/pkg/api/errors"` and then import `"errors"` as "errors".)
2. Import `"errors"`.
3. Update `deleteIngressController` to use `%w` so that its error return value can be unwrapped:

```go
func deleteIngressController(t *testing.T, cl client.Client, ic *operatorv1.IngressController, timeout time.Duration) error {
	t.Helper()
	name := types.NamespacedName{Namespace: ic.Namespace, Name: ic.Name}
	if err := cl.Delete(context.TODO(), ic); err != nil {
		return fmt.Errorf("failed to delete ingresscontroller: %w", err)
	}
	// [existing logic]
```

4. Update `assertIngressControllerDeleted` to unwrap the error value from `deleteIngressController` and return without calling `t.Fatalf` if the unwrapped error value is a not-found error.

```go
func assertIngressControllerDeleted(t *testing.T, cl client.Client, ing *operatorv1.IngressController) {
	t.Helper()
	if err := deleteIngressController(t, cl, ing, 2*time.Minute); err != nil {
		if apierrors.IsNotFound(errors.Unwrap(err)) {
			return
		}
		t.Fatalf("WARNING: cloud resources may have been leaked! failed to delete ingresscontroller %s: %v", ing.Name, err)
	// [existing logic]
```

5. Use your fancy new `assertIngressControllerDeleted` here:

```suggestion
	if err := kclient.Create(context.TODO(), ic); err != nil {
		t.Fatalf("failed to create ingresscontroller: %v", err)
	}
	defer assertIngressControllerDeleted(t, kclient, ic)
```

Up to you which approach you prefer.

Reply:
Sure, so I did call on line 36:
```
// Don't use assertIngressControllerDeleted because it's okay if it's not deleted
defer deleteIngressController(t, kclient, ic, 2*time.Minute)
```

However, I do agree with the update that it will make assertIngressControllerDeleted. TIL unwrapping errors, that's pretty cool. I'll make that update.

Done.